readers/apiserver/watchlist/watchlist.go

```diff
@@ -1,251 +1,639 @@
+		},
 	}
+}
 
-	r.mu.Lock()
-	defer r.mu.Unlock()
+func (r *Reader) createServicesWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.CoreV1().Services("").Watch(ctx, options)
+			if err != nil {
+				panic(err.Error())
+			}
+			return wi, nil
+		},
+	}
+}
+
+func (r *Reader) createDeploymentsWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.AppsV1().Deployments("").Watch(ctx, options)
+			if err != nil {
+				panic(err.Error())
+			}
+			return wi, nil
+		},
+	}
+}
 
-	newReader, err := r.newReader()
+func (r *Reader) createIngressesWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.NetworkingV1().Ingresses("").Watch(ctx, options)
+			if err != nil {
+				panic(err.Error())
+			}
+			return wi, nil
+		},
+	}
+}
+
+func (r *Reader) createEndpointsWatcher(ctx context.Context) []spawnWatcher {
+	return []spawnWatcher{
+		func(options metav1.ListOptions) (watch.Interface, error) {
+			wi, err := r.clientset.CoreV1().Endpoints("").Watch(ctx, options)
+			if err != nil {
+				panic(err.Error())
+			}
+			return wi, nil
+		},
+	}
+}
+
+// setupFilter sets up the filter cache for the Reader.
+func (r *Reader) setupFilter(ctx context.Context) error {
+	r.filterIn = make(chan watch.Event, 1)
+
+	var err error
+	r.filter, err = filter.New(ctx, r.filterIn, r.ch, r.filterOpts...)
 	if err != nil {
-		return err
+		return fmt.Errorf("error creating cache: %v", err)
 	}
+	return nil
+}
 
-	if err := newReader.SetOut(context.Background(), r.ch); err != nil {
-		return fmt.Errorf("bug: could not call SetOut() on new watchlist reader: %w", err)
+// watch watches a resource and sends the events to the cache. If this returns
+// an error, the initial watcher could not be created. This will return nil
+// if the initial watcher is created, but underlying calls go on in a goroutine.
+// This will handle automatic reconnection.
+func (r *Reader) watch(ctx context.Context, rt types.Retrieve, spawnWatchers []spawnWatcher) error {
+	if r.fakeWatch != nil {
+		return r.fakeWatch(ctx, rt, spawnWatchers)
 	}
-	if err := newReader.Run(context.Background()); err != nil {
-		return err
+
+	if ctx.Err() != nil {
+		return nil
 	}
-	if err := r.r.Close(context.Background()); err != nil {
-		r.logger.Warn(fmt.Sprintf("could not close old watchlist reader: %v", err))
+
+	var rv string
+	options := metav1.ListOptions{
+		ResourceVersion: rv,
+		Watch:           true,
 	}
-	r.r = newReader
+
+	var err error
+	watchers := make([]watch.Interface, len(spawnWatchers))
+	defer func() {
+		if err != nil {
+			for _, watcher := range watchers {
+				if watcher != nil {
+					watcher.Stop()
+				}
+			}
+		}
+	}()
+	for i := 0; i < len(spawnWatchers); i++ {
```

Comment:
pretty sure we can 

for i := range spawnWatchers {

here

Reply:
You can, but that is a new feature and I like the old ways.